File: app/services/drafting/context_gatherer.rb
# frozen_string_literal: true

module Drafting
  # Gathers related email context to inform draft responses.
  # Uses LLM to generate Gmail search queries, then fetches related threads.
  class ContextGatherer
    MAX_CONTEXT_THREADS = 3
    MAX_QUERIES = 3

    def initialize(user:, gmail_client: nil, llm_gateway: nil)
      @user = user
      @gmail_client = gmail_client || Gmail::Client.new(user: user)
      @llm = llm_gateway || LlmGateway.new
    end

    # Gather related context for an email.
    # Returns a formatted context string or nil if no relevant context found.
    def gather(sender_email:, subject:, body_text:, gmail_thread_id:)
      queries = generate_search_queries(sender_email, subject, body_text)
      return nil if queries.empty?

      threads = fetch_related_threads(queries, exclude_thread_id: gmail_thread_id)
      return nil if threads.empty?

      format_context(threads)
    rescue LlmGateway::LlmError, Gmail::Client::GmailApiError => e
      Rails.logger.warn("Context gathering failed: #{e.message}")
      nil
    end

    private

    def generate_search_queries(sender_email, subject, body_text)
      prompt = <<~PROMPT
        Based on this email, generate up to #{MAX_QUERIES} Gmail search queries to find related conversations in the user's mailbox.
        Focus on finding threads that provide context for writing a reply.

        From: #{sender_email}
        Subject: #{subject}
        Body: #{body_text&.truncate(1000)}

        Return a JSON array of search query strings. Example:
        ["from:#{sender_email} subject:project", "subject:budget Q1", "from:#{sender_email}"]

        Only return the JSON array, nothing else.
      PROMPT

      result = @llm.chat_json(
        model: AppConfig.llm.context_model,
        messages: [{ role: "user", content: prompt }],
        max_tokens: 256,
        temperature: 0.3,
        user: @user,
        call_type: "context"
      )

      queries = result[:parsed_response]
      queries.is_a?(Array) ? queries.first(MAX_QUERIES).select(&:present?) : []
    rescue LlmGateway::LlmError
      # Fallback: simple query by sender
      ["from:#{sender_email}"]
    end

    def fetch_related_threads(queries, exclude_thread_id:)
      seen_thread_ids = Set.new([exclude_thread_id])
      threads = []

      queries.each do |query|
        break if threads.length >= MAX_CONTEXT_THREADS

        begin
          response = @gmail_client.list_messages(query: query, max_results: 5)
          next unless response.messages

          response.messages.each do |msg_ref|
            break if threads.length >= MAX_CONTEXT_THREADS

            message = @gmail_client.get_message(msg_ref.id, format: "metadata")
            thread_id = message.thread_id

            next if seen_thread_ids.include?(thread_id)

            seen_thread_ids.add(thread_id)
            threads << extract_thread_summary(message)
          end
        rescue Gmail::Client::GmailApiError => e
          Rails.logger.debug("Context query failed: #{e.message}")
          next
        end
      end

      threads
    end

    def extract_thread_summary(message)
      parsed = Gmail::MessageParser.parse(message)
      {
        subject: parsed[:subject] || "(no subject)",
        sender: parsed[:sender_email],
        snippet: parsed[:snippet] || message.snippet || ""
      }
    end

    def format_context(threads)
      lines = ["Related context from your mailbox:"]
      threads.each do |thread|
        lines << ""
        lines << "- [Thread: \"#{thread[:subject]}\"]"
        lines << "  From: #{thread[:sender]}"
        lines << "  #{thread[:snippet].truncate(200)}"
      end
      lines.join("\n")
    end
  end
end

File: src/services/drafting/context-gatherer.ts
// Context gathering system - finds related email threads for draft context
import { generateContextQueries } from '../llm/service';
import type { EmailMetadata } from '../llm/types';

export interface GmailThread {
  id: string;
  subject: string;
  snippet: string;
  messages?: Array<{
    from: string;
    date: string;
    body: string;
  }>;
}

export interface ContextGatherInput {
  email: EmailMetadata;
  gmailSearch: (query: string) => Promise<GmailThread[]>;
  userId?: number;
}

export interface ContextResult {
  queries: string[];
  threads: GmailThread[];
  formattedContext: string[];
}

/**
 * Gather related email context using LLM-generated Gmail queries
 */
export async function gatherRelatedContext(
  input: ContextGatherInput
): Promise<ContextResult> {
  // Generate Gmail search queries using LLM
  const queries = await generateContextQueries({ email: input.email });

  // Execute each query and collect results
  const allThreads: GmailThread[] = [];
  for (const query of queries) {
    try {
      const threads = await input.gmailSearch(query);
      allThreads.push(...threads);
    } catch (error) {
      console.warn(`Gmail search failed for query "${query}":`, error);
    }
  }

  // Deduplicate by thread ID and limit to 3
  const uniqueThreads = deduplicateThreads(allThreads).slice(0, 3);

  // Format as context blocks
  const formattedContext = uniqueThreads.map(formatThreadAsContext);

  return {
    queries,
    threads: uniqueThreads,
    formattedContext,
  };
}

/**
 * Deduplicate threads by ID
 */
function deduplicateThreads(threads: GmailThread[]): GmailThread[] {
  const seen = new Set<string>();
  const unique: GmailThread[] = [];

  for (const thread of threads) {
    if (!seen.has(thread.id)) {
      seen.add(thread.id);
      unique.push(thread);
    }
  }

  return unique;
}

/**
 * Format thread as context block for prompt
 */
function formatThreadAsContext(thread: GmailThread): string {
  let context = `[Thread: "${thread.subject}"]\n`;

  if (thread.messages && thread.messages.length > 0) {
    // Include up to 2 most recent messages
    const recentMessages = thread.messages.slice(-2);
    for (const msg of recentMessages) {
      context += `From: ${msg.from} (${msg.date})\n`;
      context += `${msg.body.substring(0, 300)}${msg.body.length > 300 ? '...' : ''}\n`;
    }
  } else if (thread.snippet) {
    context += thread.snippet;
  }

  return context;
}
